### PR TITLE
[JUJU-1548] Default series for charms

### PR DESCRIPTION
**Pull request by hmlanigan**
Model config's default series will now act the same as using the series flag to deploy a charm. 

Historically default series in model config has been set by juju and maybe changed by users. Unfortunately there is no way to determine on pre 2.9.35 controllers who set the value. To ensure that we have a user defined value, the default series value will be cleared upon upgrade to juju 2.9.35 or migration to the same.

A follow on PR will cover this behavior for bundles.

## QA steps
Try repro steps from bug as well.

```sh
$ juju bootstrap localhost testing
# no output from the model-config command next:
$ juju model-config default-series

# deploy a charm with no default-series nor series
$ juju deploy ubuntu --dry-run 
"ubuntu" from charm-hub charm "ubuntu", revision 21 in channel stable on focal would be deployed

# Set default-series and verify used to deploy both charmhub and charmstore charms.
$ juju model-config default-series=bionic
$ juju deploy ubuntu 
Located charm "ubuntu" in charm-hub, revision 21
Deploying "ubuntu" from charm-hub charm "ubuntu", revision 21 in channel stable on bionic
$ juju deploy cs:ubuntu csubuntu
Located charm "ubuntu" in charm-store, revision 21
Deploying "csubuntu" from charm-store charm "ubuntu", revision 21 in channel stable on bionic

# Verify that the series flag takes priority
$ juju deploy ubuntu --series jammy ubuntu-jammy
Located charm "ubuntu" in charm-hub, revision 21
Deploying "ubuntu-jammy" from charm-hub charm "ubuntu", revision 21 in channel stable on jammy

# bootstrap a 2.8 controller to test migration and upgrade.
$ juju bootstrap localhost two-eight
$ juju add-model moveme
$ juju model-config default-series
bionic
$ juju model-config default-series=focal
$ juju migrate moveme testing
# wait
$ juju switch testing:moveme
$ juju model-config default-series 
bionic

$ juju switch two-eight:default
$ juju upgrade-controller --build-agent
$ juju upgrade-juju
$ juju model-config default-series
$ 
```

## Documentation changes

We are clearing the default-series and it will now be treated the same as the series flag for deploy

## Bug reference

https://bugs.launchpad.net/juju/+bug/1966664

### CODE REVIEW

File: state/upgrades.go [hunk 1]
```diff
@@ -4581,3 +4581,40 @@ func CharmOriginChannelMustHaveTrack(pool *StatePool) error {
...
+// they want. Previously it was set by juju.
+func RemoveDefaultSeriesFromModelConfig(pool *StatePool) error {
+	st, err := pool.SystemState()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var ops []txn.Op
+	coll, closer := st.db().GetRawCollection(settingsC)
+	defer closer()
+	pattern := fmt.Sprintf(":%s$", modelGlobalKey)
```
Comment (cderici) [inline state/upgrades.go:4596]: I never fully understood why we have this "e" (`modelGlobalKey`).

File: cmd/juju/application/deployer/series_selector.go [hunk 2]
```diff
@@ -75,14 +77,7 @@ func (s seriesSelector) charmSeries() (selectedSeries string, err error) {
 	// No series explicitly requested by the user.
 	// Use model default series, if explicitly set and supported by the charm.
 	if defaultSeries, explicit := s.conf.DefaultSeries(); explicit {
-		if _, err := charm.SeriesForCharm(defaultSeries, s.supportedSeries); err == nil {
-			// validate the series we get from the charm
-			if err := s.validateSeries(defaultSeries); err != nil {
-				return "", err
-			}
-			logger.Infof(msgDefaultModelSeries, defaultSeries)
```
Comment (cderici) [inline cmd/juju/application/deployer/series_selector.go:83]: The `msgDefaultModelSeries` at line 17 is dead now.

**Review by cderici — APPROVED** [review]
This LGTM. Modulo, looks like some tests need to be updated, e.g. failing with things like `no settings found for "other-application"`, and some err message mismatches.

File: state/upgrades.go (hunk 1 above)
Comment (hmlanigan) [reply state/upgrades.go:4596]: @cderici My best guess would be a hold over from juju 1.x, where models didn't exist and we had 'environ'?

File: state/migration_import.go [hunk 3]
```diff
@@ -243,6 +243,31 @@ func (ctrl *Controller) Import(model description.Model) (_ *Model, _ *State, err
...
+	v, ok := attrs["agent-version"].(string)
+	if !ok {
+		return nil, errors.New("model config missing agent-version")
+	}
+	toolsVersion, err := version.Parse(v)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// Using MustParse as the value parsed will never change.
+	newer := version.MustParse("2.9.35")
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply state/migration_import.go:263]: I completely agree. Open to other ideas.

**Review by benhoyt — APPROVED** [review]
Thanks -- nice work! Looks good and my basic QA tests worked as advertised.

**Comment by benhoyt** [comment]
The smoke and upgrade (lxd) tests failed, though those failures don't look related to this change.

File: state/upgrades.go [hunk 4]
```diff
@@ -1635,7 +1635,7 @@ func CreateMissingApplicationConfig(pool *StatePool) error {
 		ID string `bson:"_id"`
 	}
 	err = settingsColl.Find(bson.M{
-		"_id": bson.M{"$regex": bson.RegEx{"#application$", ""}}}).All(&applicationConfigIDs)
+		"_id": bson.M{"$regex": bson.RegEx{"application$", ""}}}).All(&applicationConfigIDs)
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply state/upgrades.go:1638]: Thanks for the catch! Not an intended change.

**Pull request merged**